### PR TITLE
Follow standard logging patterns and expect proper logger

### PR DESCRIPTION
**Pull request by yahehe**
This feels right to me (rejecting an improper logger, logging to console.log by default) but I'm not used to JS logging conventions.
My general idea for what to level to log at was:
If an error is thrown/returned, that's a debug
If it doesn't seem like a debug but doesn't seem like an error/warning, it's info
If it's an error that this library is swallowing and doesn't seem debug worthy then log as error.

By default this all logs to console.log, should errors log to console.err maybe?

### CODE REVIEW

File: src/auth0V1.js [hunk 1]
```diff
@@ -15,13 +15,13 @@ const saveV1TokenInCache = (config, token) => {
   const cacheKey = generateAuthV1TokenCacheKey(config);
   const decodedToken = jwt.decode(token);
   return cache.set(cacheKey, token, decodedToken.exp - decodedToken.iat)
-    .catch(err => logger(`Error saving v1 token ${cacheKey} in cache: ${JSON.stringify(err)}`));
+    .catch(err => logger.debug(`Error saving v1 token ${cacheKey} in cache: ${JSON.stringify(err)}`));
```
Comment (tocco934) [inline src/auth0V1.js]: I think this should be console.error

File: src/auth0V1.js [hunk 2]
```diff
@@ -15,13 +15,13 @@ const saveV1TokenInCache = (config, token) => {
...
 };
 
 const retrieveV1TokenFromCache = (config) => {
   const cacheKey = generateAuthV1TokenCacheKey(config);
   return cache.get(cacheKey)
-    .catch(err => logger(`Error retrieving auth0v1 token ${cacheKey} from cache with target id: ${config.targetId}: ${JSON.stringify(err)}`));
+    .catch(err => logger.debug(`Error retrieving auth0v1 token ${cacheKey} from cache with target id: ${config.targetId}: ${JSON.stringify(err)}`));
```
Comment (tocco934) [inline src/auth0V1.js]: Same here .error

File: src/auth0V1.js [hunk 3]
```diff
@@ -42,7 +42,7 @@ const retrieveV1TokenFromServer = (config) => {
   };
 
   return request(delegationOptions).catch((error) => {
-    logger(`Error retrieving auth0v1 token from Auth0 server: ${error}`);
+    logger.debug(`Error retrieving auth0v1 token from Auth0 server: ${error}`);
```
Comment (tocco934) [inline src/auth0V1.js]: .error

File: src/auth0V1.js [hunk 4]
```diff
@@ -94,7 +94,7 @@ const makeUnauthenticatedRequest = (options) => {
   });
 
   return request(noAuthRequestOptions).catch((error) => {
-    logger(`Request to ${noAuthRequestOptions.uri} failed due to: ${error}`);
+    logger.debug(`Request to ${noAuthRequestOptions.uri} failed due to: ${error}`);
```
Comment (tocco934) [inline src/auth0V1.js]: error

File: src/auth0V2.js [hunk 5]
```diff
@@ -4,15 +4,15 @@ const request = require('request-promise');
...
+let logger = require('./defaultLogger');
 
 const generateAuthV2TokenCacheKey = (config, audience) => `${audience}-${config.clientId}`;
 
 const retrieveV2TokenFromCache = (config, audience) => {
   const cacheKey = generateAuthV2TokenCacheKey(config, audience);
   return cache.get(cacheKey)
     .catch((err) => {
-      logger(`Error retrieving auth0v2 token ${cacheKey} from cache with audience ${audience}: ${JSON.stringify(err)}`);
+      logger.info(`Error retrieving auth0v2 token ${cacheKey} from cache with audience ${audience}: ${JSON.stringify(err)}`);
```
Comment (tocco934) [inline src/auth0V2.js]: error

File: src/auth0V2.js [hunk 6]
```diff
@@ -21,7 +21,7 @@ const saveV2TokenInCache = (config, audience, token) => {
   const decodedToken = jwt.decode(token);
   return cache.set(cacheKey, token, decodedToken.exp - decodedToken.iat)
     .catch((error) => {
-      logger(`Error saving v2 token ${cacheKey} in cache: ${JSON.stringify(error)}`);
+      logger.info(`Error saving v2 token ${cacheKey} in cache: ${JSON.stringify(error)}`);
```
Comment (tocco934) [inline src/auth0V2.js]: error

If we end up disagreeing with error you used .debug above

File: src/auth0V1.js [hunk 7]
```diff
@@ -110,7 +110,7 @@ const auth0v1 = (options, retryLoop, res) => {
       return Promise.reject('Not enough information for a delegation call');
     }
 
-    logger('No v1 auth possible. Attempting an unauthenticated request');
+    logger.debug('No v1 auth possible. Attempting an unauthenticated request');
```
Comment (tocco934) [inline src/auth0V1.js]: maybe .info ?

File: src/passedInAuth.js [hunk 8]
```diff
@@ -34,7 +34,7 @@ const passedInAuth = (options, retryLoop) => {
       });
     });
   }
-  logger('No token passed in. Falling back to v2.');
+  logger.debug('No token passed in. Falling back to v2.');
```
Comment (tocco934) [inline src/passedInAuth.js]: .info

**Comment by tocco934** [comment]
Sorry for the spam, I'm gonna stop since all the files are basically the same. I wasn't able to find any good guidance/practices online so my proposal is:

- .error: Use when going to throw an error or return a rejected promise
- .warn: Use when an error was returned but the library can still proceed forward
- .info: Kinda seems to fit the same purpose as .debug. Probably should just use debug when we'd want to use .info
- .debug: Helpful messages like "falling back to v2". In newer chrome windows it's hidden by default. Not sure if we should take that into account or what it would mean.
- .log: Everything else, probably shouldn't use unless we're logging requests/responses

**Comment by yahehe** [comment]
This is a library though, we shouldn't be spamming log messages to its clients. Now that I think about it probably all of this should be debug

File: src/defaultLogger.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
```
Comment (benjamingill) [inline src/defaultLogger.js]: nb: i do believe the default console has error(), warn(), info(), and debug() methods, not sure if we want to use those or not.

Comment (yahehe) [inline src/defaultLogger.js]: Yep you're right, will change

**Comment by tocco934** [comment]
Hmm good point @yahehe , it should just use debug

File: src/auth0V2.js (hunk 5 above)
Comment (tocco934) [inline src/auth0V2.js]: Ignore this, I want to see what starting a review does

Comment (tocco934) [inline src/auth0V2.js]: What? What is this?

File: src/defaultLogger.js (hunk 9 above)
Comment (yahehe) [inline src/defaultLogger.js]: Just kidding, the console doesn't have a debug method, gotta roll my own

Comment (tocco934) [inline src/defaultLogger.js]: It doesn't? Could just use .info ?

**Comment by tocco934** [comment]
You gotta pull/fetch from master & rebase. There is no fork syncing.

Do you know if there's a way to automatically update pull-requests once other pull-requests have been merged?

**Comment by yahehe** [comment]
rebased. I do not know if github will do that for us

**Pull request merged**